Error adding object with an openarray field to seq of base type

I am new to nim, sorry if this issue is inappropriate. While trying to add an openarray to an object, I encountered a c compiler error. 

The following code:
```nim
type X = ref object of RootObj
type Y = ref object of X
  field: openarray[int]

var s: seq[X]
proc f() =
  s.add(Y(field: [1]))

f()
```
Results in
```
/home/user/.cache/nim/test_d/@mtest.nim.c: In function ‘f__9bYThM6NliRXN1SXFwqXl9aw’:
/home/user/.cache/nim/test_d/@mtest.nim.c:415:2: error: cannot convert to a pointer type
  415 |  nimCopyMem((void*)(*T1_).field, (NIM_CONST void*)TM__wUwZNZbM8FcO0T0cAKGahQ_4, sizeof(tyArray__GAcaPw08Gj9cXYhAaMMr9cfw));
      |  ^~~~~~~~~~
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I'/home/user/.choosenim/toolchains/nim-#devel/lib' -I/home/user/saved/devel/games/expelled-rl/nim -o /home/user/.cache/nim/test_d/@mtest.nim.c.o /home/user/.cache/nim/test_d/@mtest.nim.c' failed with exit code: 1
```
If the sequence is not on the base class, the compiler spits an (not very explicit) error:
Error: invalid type: 'openArray[int]' in this context: 'seq[Y]' for var

### Additional Information
Tested on nim 1.4.2 and #devel dbff2cd938b326279de0f3f97b2dd8c54a90468a

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -1497,8 +1497,7 @@ proc handleConstExpr(p: BProc, n: PNode, d: var TLoc): bool =
 
 
 proc genFieldObjConstr(p: BProc; ty: PType; useTemp, isRef: bool; nField, val, check: PNode; d: var TLoc; r: Rope; info: TLineInfo) =
-  var tmp2: TLoc = default(TLoc)
-  tmp2.r = r
+  var tmp2 = TLoc(r: r)
   let field = lookupFieldAgain(p, ty, nField.sym, tmp2.r)
   if field.loc.r == "": fillObjectFields(p.module, ty)
   if field.loc.r == "": internalError(p.config, info, "genFieldObjConstr")
@@ -1513,7 +1512,12 @@ proc genFieldObjConstr(p: BProc; ty: PType; useTemp, isRef: bool; nField, val, c
     tmp2.k = d.k
     tmp2.storage = if isRef: OnHeap else: d.storage
   tmp2.lode = val
-  expr(p, val, tmp2)
+  if nField.typ.skipTypes(abstractVar).kind in {tyOpenArray, tyVarargs}:
+    var tmp3 = getTemp(p, val.typ)
+    expr(p, val, tmp3)
+    genOpenArrayConv(p, tmp2, tmp3, {})
+  else:
+    expr(p, val, tmp2)
 
 proc genObjConstr(p: BProc, e: PNode, d: var TLoc) =
   # inheritance in C++ does not allow struct initialization so

--- a/tests/views/tviews2.nim
+++ b/tests/views/tviews2.nim
@@ -35,3 +35,14 @@ block: # bug #15778
   doAssert @(reader.read(3)) == @['l', 'l', 'o']
   doAssert count == 2
 
+block: # bug #16671
+  type X = ref object of RootObj
+  type Y = ref object of X
+    field: openArray[int]
+
+  var s: seq[X]
+  proc f() =
+    s.add(Y(field: [1]))
+
+  f()
+

--- a/tests/views/tviews2.nim
+++ b/tests/views/tviews2.nim
@@ -36,13 +36,24 @@ block: # bug #15778
   doAssert count == 2
 
 block: # bug #16671
-  type X = ref object of RootObj
-  type Y = ref object of X
-    field: openArray[int]
+  block:
+    type X = ref object of RootObj
+    type Y = ref object of X
+      field: openArray[int]
 
-  var s: seq[X]
-  proc f() =
-    s.add(Y(field: [1]))
+    var s: seq[X]
+    proc f() =
+      s.add(Y(field: [1]))
 
-  f()
+    f()
 
+  block:
+    type X = ref object of RootObj
+    type Y = ref object of X
+      field: openArray[int]
+
+    var s: seq[X]
+    proc f() =
+      s.add(Y(field: toOpenArray([1, 2, 3], 0, 1)))
+
+    f()

